pkg/plugins/pg/destination/destination.go

```diff
@@ -141,24 +205,72 @@ func (s *Destination) Write(ctx context.Context, r record.Record) (record.Positi
 	// we have to manually append a semi colon to the upsert sql;
 	upsertQuery += ";"
 
-	// prepare SQL to insert cols and args into the appropriate table.
-	// suffix sql and args for upsert behavior.
-	query, args, err := psql.Insert(table).
+	colArgs, valArgs := formatColumnsAndValues(key, payload)
+
+	query, args, err := psql.
+		Insert(tableName).
 		Columns(colArgs...).
 		Values(valArgs...).
 		SuffixExpr(sq.Expr(upsertQuery)).
 		ToSql()
 	if err != nil {
-		return s.Position, cerrors.Errorf("error formatting query: %w", err)
+		return "", nil, fmt.Errorf("error formatting query: %w", err)
 	}
 
-	// attempt to run the query
-	_, err = s.db.Exec(query, args...)
-	if err != nil {
-		// return current position that hasn't been updated and the error
-		return s.Position, cerrors.Errorf("insert exec failed: %w", err)
+	return query, args, nil
+}
+
+// formatColumnsAndValues turns the key and payload into a slice of ordered
+// columns and values for upserting into Postgres.
+func formatColumnsAndValues(key, payload sdk.StructuredData) ([]string, []interface{}) {
+	var colArgs []string
+	var valArgs []interface{}
+
+	// range over both the key and payload values in order to format the
+	// query for args and values in proper order
+	for key, val := range key {
+		colArgs = append(colArgs, key)
+		valArgs = append(valArgs, val)
+		delete(payload, key) // NB: Delete Key from payload arguments
```

Comment:
Because both the key and the payload will get turned into a query column and value, so the key is still being stored in the db, we just don't want to duplicate the query column and value because Postgres will error.

Reply:
Sounds reasonable to me. 👍 I think it would still be nice that we don't mutate the input args, and simply merge the key and payload as maps and that would naturally solve the problem of duplicate entries.